docs/Workshop.md

```diff
@@ -774,3 +774,644 @@ git reset --hard domain-modeling
+
+```scala
+// here Labelable is a context bound on A, and Label is a Sum type
+enum Label:
+  case Sold, PendingSale, Undefined
+
+trait Labelable[A]:
+  def toLabel(a: A): Label
+
+def foo[A: Labelable](a: A): Label
+```
+
+It's important to note that we don't have to express a context bound
+in the generic parameters of a method in Scala 3. We can also express
+it as a `using` parameter, a parameter to a `ContextFunction`, or a
+normal parameter. All of the following mean the same thing.
+
+```scala
+def fooUsing[A](a: A)(using Labelable[A]): Label
+def fooContextFunction[A](a: A): Labelable[A] ?=> Label
+def fooNormal[A](a: A, labelable: Labelable[A]): Label
+val fooVal: [A] => Labelable[A] ?=> Label = [A] => Labelable[A] ?=> (a: A) => summon[Labelable[A]].toLabel(a)
+```
+
+In FP, we like to delay side effects until the latest possible
+moment. We do this, in general, by encoding programs as data types and
+defininig an interpreter that converts a datatype into an executable
+program at runtime. This can be a little tedious without helper
+interfaces. Imagine programming `fooVal` above like this:
+
+```scala
+Apply(
+  Select(
+    Apply(
+	  TypeApply(
+	    ClassOf(
+		  Summon.class
+	   ), 
+	   List(
+	     TypeTree(
+		   TypeApply(
+		     ClassOf(
+			   Labelable.class
+		     ),
+			 List(
+			   TypeParamRef("A")
+		     )
+		   )
+	     )
+	   )
+	  ), 
+	  List()
+	), 
+    "toLabel"
+  ), 
+  List(
+    ParamRef("a")
+  )
+)
+```
+
+That is clearly not very ergonomic. `tagless final` allows us to use
+all the complexity limiting tricks we talked about for values to
+encode programs with small, parameterized methods which produce trees
+of data values encoding a given, possibly side-effecting domain, like
+say handling http requests and responses behind the implementations of
+the interfaces. The data types that are built then expose some sort of
+`run` method, which interprets the value of the tree produced.
+
+In this way, programs written with a tagless final encoding limit the
+cumbersome data type encodings required to work with side-effecting
+programs while still allowing users to reason about the code as a tree
+of immutable values. This allows the programmer to refactor tagless
+final programs safely.
+
+It allows the designers of libraries to control and optimize the
+interpretation of the trees produced by effectful programs without
+adding additional complexity to the interfaces used by users of these
+libraries. Method calls to the interfaces are free to modify the input
+trees of previous calls, for example.
+
+Meanwhile, users are prevented from violating the calling conventions
+of the library by coding against the interface rather than the
+implementation, enforced by the typechecker. The underlying complexity
+is fully abstracted away from the user.
+
+New methods can be added to a library without violating the source
+compatibility of code built against earlier versions. And many other
+advantages.
```

Comment:
😲  I have been using tagless final for a while, and I will honestly say I never reasoned about explaining in that way. Don't you think it will too complex/abstract for some students?

Reply:
We're going to hide it behind a collapsed detail link in a soon to come PR.

Some of the explanation comes from comments made by Fabio Labella in an [old typelevel theory chat](https://discord.com/channels/632277896739946517/839263668478672937/847438971969077248):

Particularly 

* [encoding an algebra as functions](https://discord.com/channels/632277896739946517/839263668478672937/847446152487108608)
* The [separation of concerns between interpreters and algebras](https://discord.com/channels/632277896739946517/839263668478672937/847447910853640222)
* The [doesn't need to be implicit parameters](https://discord.com/channels/632277896739946517/839263668478672937/847449140778762280)
* The mixing of [free and tagless final encodings for user-level features and implementers](https://discord.com/channels/632277896739946517/839263668478672937/847453412915544134)
* And the fact that effect interpreters are ["tagged"](https://discord.com/channels/632277896739946517/839263668478672937/847456529648582697)
* And the [general advantages](https://discord.com/channels/632277896739946517/839263668478672937/847457565230891048)

, some of it comes from "Constraints Liberate, Liberties Constrain" by Runar Bjarnsson (working with too large of [interfaces](https://www.youtube.com/watch?v=GqmsQeSzMdw&t=323s) and [generic interfaces](https://www.youtube.com/watch?v=GqmsQeSzMdw&t=490s)) and Sam Halliday's Fp in Scala for Mortals ([Interface Sizes (4.1.7)](https://leanpub.com/fpmortals/read#leanpub-auto-data), [Duality between Free and TF encodings and ASTs](https://leanpub.com/fpmortals/read#leanpub-auto-a-free-lunch)). 

Some of it comes from how I've explained it as small interfaces in the past without actually going into the actual theory and papers behind it.